clash-lib/src/Clash/Core/Evaluator/Semantics.hs

```diff
@@ -167,20 +168,40 @@ evaluateVarWith eval i
     Just x  -> forceLocal x
     Nothing -> pure (VNeu (NeVar i))
    where
+    -- TODO Does this need to change?
     forceLocal = either (withoutLocal i . eval) pure
 
   goGlobal :: Eval Value
-  goGlobal = getGlobal i >>= \case
-    Just b
-      | bindingSpec b /= NoInline -> do
-          v <- forceGlobal (bindingTerm b)
-
-          updateGlobal i v
-          pure v
-
-    _ -> pure (VNeu (NeVar i))
+  goGlobal = do
+    gRecInfo <- getRecInfo
+    gFuel <- getFuel
+
+    getGlobal i >>= \case
+      Just b
+        -- The binding can't be inlined.
+        |  bindingSpec b == NoInline
+        -> pure (VNeu (NeVar i))
+
+        -- The binding can be inlined if there is enough fuel.
+        |  isRecursive i gRecInfo
+        -> if gFuel == 0
+             then pure (VNeu (NeVar i))
+             else do
+               v <- forceGlobal (bindingTerm b)
+               putFuel (gFuel - 1)
+               updateGlobal i v
+               pure v
```

Comment:
Should fuel be kept per binding instead? I don't have strong feelings either way, but I'm wondering about the reasoning (easy to implement being a valid reason :-)).

Reply:
Or at least per recursive group. Although it becomes problematic when we evaluate a binding from one recursive group inside the binding of another recursive group.

Reply:
I did initially think of per-binder / per-group but went with this because (1) it was more straightforward to implement and (2) it felt slightly easier to reason about the result, as all paths from root to leaf of the final AST use at most n fuel

A consequence is that programs with wide ASTs may incur a performance penalty because of this, but I would hope that can be spotted / tuned later